fun.lua

```diff
@@ -1032,6 +1040,86 @@ methods.operator = operator
 exports.op = operator
 methods.op = operator
 
+--------------------------------------------------------------------------------
+-- Stateful iterators
+--------------------------------------------------------------------------------
+
+-- Returns 'gen' throwing an error that explains that method is not supported
+local unsupported_stateful_method_gen = function(method)
+    return function()
+        error(string.format("stateful iterators do not support method '%s'",
+              method))
+    end
+end
+
+-- Helper of 'wrap_with_elems'
+local gen_with_elems_wrapper = function(gen_param_elems, state)
+    local elems = gen_param_elems[3]
+    if not elems then
+        return gen_param_elems[1](gen_param_elems[2], state)
+    else
+        gen_param_elems[3] = nil
+        return state, unpack(elems)
+    end
+end
+
+-- Wrap an iterator that yields all values from 'elems' table on the first
+-- iteration and then works in the same way as 'wrap(gen, param, state)'
+local wrap_with_elems = function(elems, gen, param, state)
+    local gen_param_elems = {gen, param, elems}
+    return wrap(gen_with_elems_wrapper, gen_param_elems, state)
+end
+
+-- Checks if drop_while should continue skipping. If iterator is not exhausted,
+-- elements returned by iterator are wrapped into a table and returned as the
+-- third return value
+local stateful_drop_while_check = function(fun, state_x, ...)
+    if state_x == nil then
+        return state_x, false
+    end
+    return state_x, fun(...), {...}
+end
+
+-- Implementation of method drop_while: original one copies iterator state to
+-- iterate over the same element twice - it doesn't work with stateful iterators
+local stateful_drop_while = function(fun, gen_x, param_x, state_x)
```

Comment:
I would highlight a question from our voice discussion: does the new `drop_while` worse than the original one for a stateless iterator?

If no, I would just replace the original `drop_while` with the new implementation.

We can use, for example, https://luajit.me to perform some comparisons. From my experience we should [take care](https://github.com/GitSparTV/LuaJIT-Benchmarks/issues/3) to counting instructions within a loop differently from ones outside. Also, it is desirable to verify the analysis using some synthetic data against timings on a real hardware.

Reply:
My implementation was slightly rewritten to improve performance - now it's even a bit better than the old one. So I just replaced `drop_while` with the new implementation and dropped stateful iterators completely.

Reply:
Nice!